The task is to predicct whether a bank currency note is authentic or not based on four attributes i.e.,variance of the image,wavelet transformed image,skewness,entropy and curtosis of the image.

Data Set Information:

Data were extracted from images that were taken from genuine and forged banknote-like specimens. For digitization, an industrial camera usually used for print inspection was used. The final images have 400x400 pixels. Due to the object lens and distance to the investigated object gray-scale pictures with a resolution of about 660 dpi were gained. Wavelet Transform tool were used to extract features from images.

Attribute Information:
1. variance of Wavelet Transformed image (continuous)
2. skewness of Wavelet Transformed image (continuous)
3. curtosis of Wavelet Transformed image (continuous)
4. entropy of image (continuous)
5. class (integer)

In [1]:
import pandas as pd
import numpy as np

In [2]:
dataset = pd.read_csv("bank_authentication.csv")

In [4]:
dataset.head()

Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

In [5]:
dataset.isnull().sum()

Variance    0
Skewness    0
Curtosis    0
Entropy     0
Class       0
dtype: int64

In [6]:
#Prepare the dataset for training
X = dataset.iloc[:,0:4].values
y = dataset.iloc[:,4].values

In [9]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 0)

In [10]:
#Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [23]:
#Training the algorithm
from sklearn.ensemble import RandomForestClassifier
cl = RandomForestClassifier(n_estimators=30,random_state = 1)
cl.fit(X_train,y_train)
y_predict = cl.predict(X_test)

n_estimators :integer optional (default = 100)

The number of trees in the forest

Criterion :string,optional (default = 'gini')

The function to measure the quality of the split.This parameter is tree specific

In [24]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
print(confusion_matrix(y_test,y_predict))
print(accuracy_score(y_test,y_predict))
print(classification_report(y_test,y_predict))

[[155   2]
 [  1 117]]
0.9890909090909091
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       157
           1       0.98      0.99      0.99       118

    accuracy                           0.99       275
   macro avg       0.99      0.99      0.99       275
weighted avg       0.99      0.99      0.99       275



#Unfortunately there's no easy to graph the "best" tree or an overall ensemble tree from 
your forest,just a random example u need to consider a specific estimator.

In [29]:
from sklearn.externals.six import StringIO
from sklearn.tree import export_graphviz
from  pydotplus import graph_from_dot_data
from IPython.display import Image
dot_data = StringIO()
export_graphviz(cl.estimators_[1],out_file=dot_data,
                filled = True,rounded = True,
                class_names=['0','1'])
graph = graph_from_dot_data(dot_data.getvalue())
Image(graph.create_png())
graph.write_pdf('1sttree.pdf')

True

In [28]:
cl.predict([[3.6216,8.6661,-2.8073,-0.44699]])

array([0], dtype=int64)

The problem here is to predict the gas consumption (in millions of gallons) in 48 of the US states based on petrol tax (in cents), per capita income (dollars), paved highways (in miles) and the proportion of population with the driving license

In [1]:
import numpy as np
import pandas as pd
dataset = pd.read_csv('petrol_consumption.csv')

In [17]:
dataset

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.00            3571            1976                         0.525   
1         9.00            4092            1250                         0.572   
2         9.00            3865            1586                         0.580   
3         7.50            4870            2351                         0.529   
4         8.00            4399             431                         0.544   
5        10.00            5342            1333                         0.571   
6         8.00            5319           11868                         0.451   
7         8.00            5126            2138                         0.553   
8         8.00            4447            8577                         0.529   
9         7.00            4512            8507                         0.552   
10        8.00            4391            5939                         0.530   
11        7.50            5126           14186                         0.525   
12        7.00            4817            6930                         0.574   
13        7.00            4207            6580                         0.545   
14        7.00            4332            8159                         0.608   
15        7.00            4318           10340                         0.586   
16        7.00            4206            8508                         0.572   
17        7.00            3718            4725                         0.540   
18        7.00            4716            5915                         0.724   
19        8.50            4341            6010                         0.677   
20        7.00            4593            7834                         0.663   
21        8.00            4983             602                         0.602   
22        9.00            4897            2449                         0.511   
23        9.00            4258            4686                         0.517   
24        8.50            4574            2619                         0.551   
25        9.00            3721            4746                         0.544   
26        8.00            3448            5399                         0.548   
27        7.50            3846            9061                         0.579   
28        8.00            4188            5975                         0.563   
29        9.00            3601            4650                         0.493   
30        7.00            3640            6905                         0.518   
31        7.00            3333            6594                         0.513   
32        8.00            3063            6524                         0.578   
33        7.50            3357            4121                         0.547   
34        8.00            3528            3495                         0.487   
35        6.58            3802            7834                         0.629   
36        5.00            4045           17782                         0.566   
37        7.00            3897            6385                         0.586   
38        8.50            3635            3274                         0.663   
39        7.00            4345            3905                         0.672   
40        7.00            4449            4639                         0.626   
41        7.00            3656            3985                         0.563   
42        7.00            4300            3635                         0.603   
43        7.00            3745            2611                         0.508   
44        6.00            5215            2302                         0.672   
45        9.00            4476            3942                         0.571   
46        7.00            4296            4083                         0.623   
47        7.00            5002            9794                         0.593   

    Petrol_Consumption  
0                  541  
1                  524  
2       

In [3]:
dataset.isnull().sum()

Petrol_tax                      0
Average_income                  0
Paved_Highways                  0
Population_Driver_licence(%)    0
Petrol_Consumption              0
dtype: int64

In [4]:
#Prepare the dataset for training
X = dataset.iloc[:, 0:4].values
y = dataset.iloc[:, 4].values

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

#Feature Scaling
We know our dataset is not yet a scaled value, for instance the Average_Income field has values 
in the range of thousands while Petrol_tax has values in range of tens. 
Therefore, it would be beneficial to scale our data 

In [9]:
#Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [10]:
#Training the Algorithm
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(n_estimators=20, random_state=0)
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)

n_estimators:integer, optional (default=10)

The number of trees in the forest.

criterion :string, optional (default=”mse”)
The function to measure the quality of a split. Supported criteria are “mse” for the mean squared error, which is equal to variance reduction as feature selection criterion, and “mae” for the mean absolute error.

In [14]:
import numpy as np
from sklearn import metrics
print('Mean Absolute Error:',metrics.mean_absolute_error(y_test,y_pred))
print('Mean Squared Error:',metrics.mean_squared_error(y_test,y_pred))
print('Root Mean Squated Error',np.sqrt(metrics.mean_squared_error(y_test,y_pred)))

Mean Absolute Error: 51.76500000000001
Mean Squared Error: 4216.166749999999
Root Mean Squated Error 64.93201637097064


With 20 trees,the root mean squared error is 64.93 which is greate than 10% of the average of petrol consumption i.e.,576.77.This indicates that we have not used enough number of estimators(trees).

In [15]:
y_pred = regressor.predict([[9.0,3571,1976,0.525]])

In [16]:
y_pred

array([506.1])

In [34]:
#Training the Algorithm
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(n_estimators=150, random_state=0)
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)

In [35]:
import numpy as np
from sklearn import metrics
print('Mean Absolute Error:',metrics.mean_absolute_error(y_test,y_pred))
print('Mean Squared Error:',metrics.mean_squared_error(y_test,y_pred))
print('Root Mean Squated Error',np.sqrt(metrics.mean_squared_error(y_test,y_pred)))

Mean Absolute Error: 48.09866666666668
Mean Squared Error: 3539.186853333334
Root Mean Squated Error 59.49106532357052


In [36]:
y_pred = regressor.predict([[7.50,4870,2351,0.529]])

In [37]:
y_pred

array([505.05333333])